src/formatting/expr.rs

```diff
@@ -214,14 +214,45 @@ pub(crate) fn format_expr(
         ast::ExprKind::AddrOf(borrow_kind, mutability, ref expr) => {
             rewrite_expr_addrof(context, borrow_kind, mutability, expr, shape)
         }
-        ast::ExprKind::Cast(ref expr, ref ty) => rewrite_pair(
-            &**expr,
-            &**ty,
-            PairParts::infix(" as "),
-            context,
-            shape,
-            SeparatorPlace::Front,
-        ),
+        ast::ExprKind::Cast(ref expr, ref ty) => {
+            /* Retrieving the comments before and after cast */
+            let i = context
+                .snippet_provider
+                .span_before(mk_sp(expr.span.hi(), ty.span.lo()), "as");
+            let infix_prefix_span = mk_sp(expr.span.hi(), i);
+            //let infix_prefix_comments = context.snippet(infix_prefix_span);
+            let c = rewrite_missing_comment(infix_prefix_span, shape, context)?;
+            let mut infix_prefix_comments = String::with_capacity(c.len() + 1);
+            if !c.is_empty() {
+                infix_prefix_comments.push(' ');
+            }
+            infix_prefix_comments.push_str(&c);
+            let i = context
+                .snippet_provider
+                .span_after(mk_sp(expr.span.hi(), ty.span.lo()), "as");
+            let infix_suffix_span = mk_sp(i, ty.span.lo());
+            //let infix_suffix_comments = context.snippet(infix_suffix_span);
+            let mut infix_suffix_comments =
+                rewrite_missing_comment(infix_suffix_span, shape, context)?;
+            if !infix_suffix_comments.is_empty() {
+                infix_suffix_comments.push(' ');
+            }
+
+            rewrite_pair(
+                &**expr,
```

Comment:
There's a few things we should change here. By renaming the enum value to something else, like `subexpr` we can simplify the process of creating the correct spans and extracting the comments from them. We also should defer to `rewrite_pair` to add the separating space when needed between the lhs & infix_prefix & infix_suffix and rhs instead of trying to do so here. That way we can avoid the extra work here of adding a space (which is then potentially peeled off), and can also encapsulate the logic of whether a separating space is needed within `rewrite_pair`, instead of each caller duplicating this logic. Have also left some inline items on `rewrite_pair` to that end

```suggestion
        ast::ExprKind::Cast(ref subexpr, ref ty) => {
           let prefix_span = mk_sp(
                subexpr.span.hi(),
                context.snippet_provider.span_before(expr.span, "as"),
            );
            let suffix_span = mk_sp(
                context.snippet_provider.span_after(expr.span, "as"),
                ty.span.lo(),
            );
            let infix_prefix_comments = rewrite_missing_comment(prefix_span, shape, context)?;
            let infix_suffix_comments = rewrite_missing_comment(suffix_span, shape, context)?;
            
            rewrite_pair(
                &**subexpr,
```